### PR TITLE
WIP: send welcome email

### PR DESCRIPTION
**Pull request by pdehaan**
Questions:
- [**sendmail**](https://www.npmjs.com/package/sendmail) vs [**nodemailer**](https://www.npmjs.com/package/nodemailer) -- It looks like nodemailer has support for different transport options (such as Amazon SES via [**nodemailer-ses-transport**](https://github.com/andris9/nodemailer-ses-transport)), plugins (DKIM), HTML+plaintext...
- When should we use SES vs just SMTP? Currently just has a big `if (config.get('env') === 'local') {...}` block, but not sure if we should use SMTP when `env` is "test".
- There has to be a better way of doing templates than my hacky thing of using `_.template` (since we were already requiring underscore in package.json). Not sure if we should be using Handlebars or something else. I need to dig into what content/auth server is doing to see if there is a more official way.
- Should we have different queue tasks for welcome emails vs other emails, or just switch statements in the server/work-queue/jobs/send-email.js to say `if (o.type === 'welcome') {...}` versus just having specific jobs like "send-welcome-email.js".

### CODE REVIEW

**Comment by nchapman** [comment]
> When should we use SES vs just SMTP

This reminds me that @mostlygeek mentioned using SES's SMTP gateway to make it easier to switch between services. Would that make things easier or harder?

**Comment by 6a68** [comment]
About the questions in the description:
- sendmail vs nodemailer: I have no idea. Take a guess; we'll iterate as we actually get towards production.
- SES vs SMTP: I like Benson's idea of always using an SMTP gateway; that should hopefully make the code simpler
- Templating: I think _.template is overkill, just do string concatenation and move along
- Other queue tasks: I think it would be fine, for now, to just call it the `send-welcome-email` job. When we add other email tasks, we'll be able to decide the right level of generality / code-sharing.

**Comment by 6a68** [comment]
btw really nice work! this looks great

File: server/work-queue/jobs/send-email.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+module.exports = {
+  // o is an object with keys { email }
+  perform: function(o, cb) {
+    var toEmail = o.email;
+    log.verbose('job created with params ' + JSON.stringify(o));
+
+    switch (config.get('env')) {
+      case 'local':
+        sendmail({
+          from: 'info@chronicle.firefox.com',
```
Comment (6a68) [inline server/work-queue/jobs/send-email.js]: I'm pretty sure we will need to vary the 'from' address. Maybe that should be a sendmail config?

File: server/work-queue/jobs/send-email.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+var contentTmpl = _.template('Welcome to Chronicle, <%= email %>!');
+
+module.exports = {
+  // o is an object with keys { email }
+  perform: function(o, cb) {
+    var toEmail = o.email;
+    log.verbose('job created with params ' + JSON.stringify(o));
+
+    switch (config.get('env')) {
```
Comment (6a68) [inline server/work-queue/jobs/send-email.js]: Let's just stick the `local` env case inline here and remove all these other cases and remove the switch statement / env check completely. We'll get there when we get there. Travis / intern don't currently exercise the new user path, so we don't have to solve the `test` env right now.

File: server/work-queue/jobs/send-email.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      case 'local':
+        sendmail({
+          from: 'info@chronicle.firefox.com',
+          to: toEmail,
+          subject: 'Welcome to Chronicle',
+          content: contentTmpl({email: toEmail})
```
Comment (6a68) [inline server/work-queue/jobs/send-email.js]: yeah, seriously, I'd just do `content: 'welcome to chronicle, ' + email + '!'`

File: server/work-queue/jobs/send-email.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        }, function (err, reply) {
+          if (err) {
+            log.error('Unable to send email: ' + err);
+            cb(err);
+            throw err;
```
Comment (6a68) [inline server/work-queue/jobs/send-email.js]: I played with the queue a bit more yesterday; I think `cb(err)` tells node-resque the job failed, and we don't need to throw. Sorry I've been unclear on this point.

**Comment by 6a68** [comment]
@pdehaan looks great, man! feel free to ping me if you have followup questions, otherwise this is fine to land once it's updated.

**Comment by 6a68** [comment]
oh wait, I guess it needs to be rebased. Ping me if that goes sideways :-P

**Comment by 6a68** [comment]
After looking a bit more, I think nodemailer might be a better bet than sendmail, only because it has way more people using it, and it's been steadily maintained. (Maybe I'm wrong, and sendmail is done in some sense, because it has very few features? Feel free to dig further and make the final call here.)

**Comment by pdehaan** [comment]
OK, attempt 2. Rip it apart.

TL;DR:
- added some configs to convict so we can customize per environment (and updated config/local.json.example -- although possibly foolishly since "local" doesn't use SES):
```js
email: {
  fromEmail: {
    format: 'email',
    default: 'info@chronicle.firefox.com'
  },
  ses: {
    accessKeyId: {
      format: String,
      default: 'your_key_here'
    },
    secretAccessKey: {
      format: String,
      default: 'your_key_here'
    }
  }
},
```
- changed from **sendmail** to **nodemailer**.
- using hypothetical AWS SES transport if `NODE_ENV` != "local".
- renamed to `send-welcome-email` and queue job name to `sendWelcomeEmail`.
- sends plaintext and equally ugly HTML welcome emails (thanks to the mighty power of the `<strong>` tag).

**Comment by mostlygeek** [comment]
I would suggest skipping the SES api and just use SMTP. If you need a local one then we can run postfix (or whatever). 

When we authenticate to SES SMTP we will use a username/password. This be different from an AWS IAM access key/pass. So you should have a smtp user/password in convict. 

And for my sanity, please have a *flat* convict schema. 

Ben says no to this: 
````
email: {
  fromEmail: {
    format: 'email',
    default: 'info@chronicle.firefox.com'
  },
  ses: {
    accessKeyId: {
      format: String,
      default: 'your_key_here'
    },
    secretAccessKey: {
      format: String,
      default: 'your_key_here'
    }
  }
},
````

Having a deeply nested schema makes it impossible for us to merge default conf + env confs + secrets into a single JSON object during deployment. Every time a recursive deep merge is required that cute science doggy gets no dinner.

File: config/local.json.example [hunk 5]
```diff
@@ -27,6 +27,13 @@
     "apiKey": "",
     "secretKey": ""
   },
+  "email": {
+    "fromEmail": "info@chronicle.firefox.com",
```
Comment (6a68) [inline config/local.json.example]: Email sent with this from address never arrived in my gmail account. I don't know what a good solution is here, though, so we can just leave it.

**Comment by 6a68** [comment]
@pdehaan looks great, works for me, feel free to rebase + merge

**Comment by 6a68** [comment]
How you ask?
  1.  `git rebase -i HEAD~3` from this branch (because we want to squash the most recent 3 commits together)
  2. for the two commits under the main commit, change 'pick' to 's' (for 'squash'):

screenshot 2015-01-30 15 01 29

  3. save and quit out of the editor
  4. you'll get bounced to a new editor session. fixup the commit message here:

starts like
screenshot 2015-01-30 15 04 14

make it something like
screenshot 2015-01-30 15 05 05

  5. save and quit out of that editor. feel your git neckbeard grow thick and mighty. rejoice.
screenshot 2015-01-30 15 05 45

**Comment by 6a68** [comment]
oh, forgot the last step: force push (git push -f) to this branch, then hit that big green button

**Comment by 6a68** [comment]
@mostlygeek we should talk about how you guys could use ruby yaml tools to handle non-flat configs. It seems kinda confusing to me that we're using puppet (a ruby tool), and we're feeding it yaml configs (the ruby conf file format), but we're not using a ruby one-liner to transform the configs. cc @nchapman

**Comment by pdehaan** [comment]
On the topic of flat configs, doesn't [**fxa-content-server**'s config](https://github.com/mozilla/fxa-content-server/blob/master/server/lib/configuration.js) use the same/similar nesting?

**Comment by pdehaan** [comment]
Anybody else care to re-re-review before I press the green button?
I ripped out the SES transporter and replaced with an SMTP transporter, at the request of @mostlygeek.

Otherwise, we can merge now and refactor the code next week.

**Comment by 6a68** [comment]
sure, I'll do a final skim

**Comment by 6a68** [comment]
lgtm, fire when ready

**Comment by 6a68** [comment]
ok, I'll hit that button ^_^
